neps/nep-0485.md

```diff
@@ -0,0 +1,102 @@
+
+The key for data in SES is the hash of the value. This makes it immutable and allows users to trust that the shared code is never modified.
+
+## Motivation
+
+- Certain basic functionality contracts such as dead man switch or multi-signature contracts require ~1 NEAR just to cover the storage staking. If a user is using zero-balance accounts and meta-transactions to cover costs, it is highly likely that they wouldn't have any NEAR on their accounts.
+
+- There are certain proxy contracts that are repeatedly deployed to different accounts for certain apps to function. One example is KeyPom, which deploys a contract costing 0.4 NEAR as a proxy contract for all of the trial accounts. This is not sustainable for businesses or individual users.
+
+As this is a part of the [Account Extensions upgrade](https://github.com/near/NEPs/issues/478), some of the benefits of account namespaces are realized in conjunction with other proposals from the upgrade:
+
+- Contract namespaces (submodules) can reference SES code, enabling easy and cheap contract composability.
+
+## Specification
+
+### The `SES` Trie Key
+
+We introduce a new trie key: `SES`. It maps code hashes to epoch IDs. This defines when the code "expires" and is updated to "current_epoch +2" every time the code is accessed.
+
+### The `SES` RocksDB Column
+
+The contract code itself lives in a separate RocksDB column, which must also be synced during state sync.
+
+### `DeployContract` action
+
+The `DeployContract` action is modified to include a flag named `is_ses` that would deploy the contract to SES. The effect of this flag is that the contract code is not stored in the account's storage, but in the `SES` column instead, and the SES code is referred to by the receiving account as its deployed contract code.
+
+The cost to deploy to SES is the same as deploying a contract normally.
+
+### RPC view calls
+
+#### `view_account`
+
+The `view_account` call now also returns a field called `shard_id` that reveals which shard the account lives on.
+
+#### `view_code`
+
+The `view_code` call returns the SES code just as it would return the code of a normally-deployed contract. A new field, `is_ses`, is also introduced, containing a boolean value indicating whether the code is stored in SES.
+
+## Reference Implementation
+
+- WIP
+
+## Security Implications
+
+- When SES code expires, it is deleted, rendering every referent soft-locked. The soft lock is resolved when any account referencing the code re-deploys it. The re-deploy must be of identical code to the same shard ID.
+
+## Drawbacks
+
+- Exposes sharding to developers.
+- Code expires.
+- More complicated `DeployContract` action.
+
+## Unresolved Issues (Optional)
+
+- How long should the code live in SES before it expires?
+- How would SES integrate with dynamic sharding?
+- What happens when you `FUNCTION_CALL` a SES contract that references expired code?
```

Comment:
Could you clarify/outline the motivation for expiring contract code (why expiring is needed, and what are the alternatives) ?

Reply:
If the code doesn't expire, there's no guarantee that its storage will be paid for, but the validators are still required to keep it alive.